### PR TITLE
simplify pulls constants out of integral

### LINKED ISSUE #15965: Integral.simplify should extract constants
Compare the extraction of `y` in one case but not the other:
```
>>> Sum(x*y, (x, 1, n)).simplify()
    n    
   __    
   \ `   
y*  )   x
   /_,   
  x = 1  
>>> Integral(x*y, (x, 1, n)).simplify()
  n       
  /       
 |        
 |  x*y dx
 |        
/         
1  
```
This extraction would help checkode recognize solutions where the constants appear inside the Integral (and will help any other simplification where constants appear in or out of the integrand within an expression). e.g #15913 

This applies to Derivatives, too, where a y could be pulled out of `Derivative(y*f(x), x)`. cf #15964

### PR DESCRIPTION
**Pull request by smichr**
#### References to other Issues or PRs

closes #15965 
closes #15968 
#### Brief description of what is fixed or changed

simplifying Integral now behaves the same as simplifying a Sum in that any constant factors will appear outside the Integral.
```python
>>> Integral(x*y, (x, 1, 3)).simplify()
y*Integral(x, (x, 1, 3))
```
#### Other comments
When the integrand is a sum, terms that are independent will also be separated from the dependent terms and be multiplied by an unevaluated integral of unity:

```python
>>> simplify(Integral(x + y, (x, 1, 3)))
y*Integral(1, (x, 1, 3)) + Integral(x, (x, 1, 3))
```

Code for Sum and Integral is condensed into `_factor_sum_int` which is called from `factor_terms` (which is called from `simplify_sum` which is called from `simplify` in the case of `Sum` or directly from `simplify` in the case of `Integral` which does not have an equivalent `simplify_integral` routine.

`factor_sum` should probably be deprecated; it is in the `import *` namespace; a direct call to `factor_terms` can be made instead. 

I am not sure if it is necessary to restrict `_factor_sum_int` to only removing multiplicative constants when called from `factor_terms`.

A caveat in using this is that if a symbol is being used temporarily to hold the place of an x-dependent integrand/summand and `factor_terms` or `simplify` is used on the expression, the *wrong* result will be obtained when backsubstituting in the function for that symbol.
```python
>>> eq = Integral(temp, x)
>>> factor_terms(eq)
temp*Integral(1, x)
>>> _.subs(temp, x).doit() == eq.subs(temp, x).doit()
False
```
#### Release Notes

* core
    * exprtools - factor_terms now factors constants out of Integrals
* simplify
    * simplify will now factor constants out of an Integral

### CODE REVIEW

**Comment by smichr** [comment]
So the problem with the failing beam tests is that (I believe) Dummy variables are used in Integrals which must represent functions dependent on the variable of integration. So when these are replaced with the actual functions they don't result in the correct integral. I haven't yet tracked down where this is happening and why Dummies are used instead of a unique function dependent on the variable of interest is used, e.g.

```python
def unique_func(eq):
    names = [f.name for f in eq.atoms(AppliedUndef)]
    for f in numbered_symbols('f'):
        if f not in names:
            return f

# need a dummy functions not in eq
f = unique_func(eq)
y = Integral(f(x)*eq, x)  # instead of Integral(Dummy()*eq, x)
```

File: sympy/physics/continuum_mechanics/beam.py [hunk 1]
```diff
@@ -1526,20 +1526,17 @@ class Beam3D(Beam):
     >>> b.solve_slope_deflection()
     >>> b.slope()
     [0, 0, l*x*(-l*q + 3*l*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(2*(A*G*l**2 + 12*E*I)) + 3*m)/(6*E*I)
-    + q*x**3/(6*E*I) + x**2*(-l*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(2*(A*G*l**2 + 12*E*I))
-    - m)/(2*E*I)]
+        + x**2*(-3*l*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(2*(A*G*l**2 + 12*E*I)) - 3*m + q*x)/(6*E*I)]
```
Comment (smichr) [inline sympy/physics/continuum_mechanics/beam.py]:
```python
>>> old=S('''l*x*(-l*q + 3*l*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(2*(A*G*l**2 + 12*E*I)) + 3*m)/(6*E*I)
...     + q*x**3/(6*E*I) + x**2*(-l*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(2*(A*G*l**2 + 12*E*I))
...     - m)/(2*E*I)''')
>>> new=S('''l*x*(-l*q + 3*l*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(2*(A*G*l**2 + 12*E*I)) + 3*m)/(6*E*I)+ x**2*(-3*l*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(2*(A*G*l**2 + 12*E*I)) - 3*m + q*x)/(6*E*I)''')
>>> old.equals(new)
True
```

File: sympy/physics/continuum_mechanics/beam.py [hunk 2]
```diff
@@ -1526,20 +1526,17 @@ class Beam3D(Beam):
...
-    >>> simplify(dy - expectedy)
-    0
+    >>> dy = collect(simplify(dy), x)
+    >>> dx == dz == 0
+    True
+    >>> dy == (x*(12*A*E*G*I*l**3*q - 24*A*E*G*I*l**2*m + 144*E**2*I**2*l*q +
+    ...           x**3*(A**2*G**2*l**2*q + 12*A*E*G*I*q) +
+    ...           x**2*(-2*A**2*G**2*l**3*q - 24*A*E*G*I*l*q - 48*A*E*G*I*m) +
+    ...           x*(A**2*G**2*l**4*q + 72*A*E*G*I*l*m - 144*E**2*I**2*q)
+    ...           )/(24*A*E*G*I*(A*G*l**2 + 12*E*I)))
```
Comment (smichr) [inline sympy/physics/continuum_mechanics/beam.py]: >>> old=S('''(
...     -l**2*q*x**2/(12*E*I) + l**2*x**2*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(8*E*I*(A*G*l**2 + 12*E*I))
...     + l*m*x**2/(4*E*I) - l*x**3*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(12*E*I*(A*G*l**2 + 12*E*I)) - m*x**3/(6*E*I)
...     + q*x**4/(24*E*I) + l*x*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(2*A*G*(A*G*l**2 + 12*E*I)) - q*x**2/(2*A*G)
...     )''')
>>> new=S('''(x*(12*A*E*G*I*l**3*q - 24*A*E*G*I*l**2*m + 144*E**2*I**2*l*q +
...                x**3*(A**2*G**2*l**2*q + 12*A*E*G*I*q) +
...                x**2*(-2*A**2*G**2*l**3*q - 24*A*E*G*I*l*q - 48*A*E*G*I*m) +
...                x*(A**2*G**2*l**4*q + 72*A*E*G*I*l*m - 144*E**2*I**2*q)
...                )/(24*A*E*G*I*(A*G*l**2 + 12*E*I)))''')
>>> old.equals(new)
True

File: sympy/solvers/ode.py [hunk 3]
```diff
@@ -4094,11 +4094,14 @@ def unreplace(eq, var):
     var = func.args[0]
     subs_eqn = replace(eq, var)
     try:
-        solns = solve(subs_eqn, func)
+        # turn off simplification to protect Integrals that have
+        # _t instead of fx in them and would otherwise factor
+        # as t_*Integral(1, x)
```
Comment (smichr) [inline sympy/solvers/ode.py]: The other option would be to use an `intx` analogous to `diffx` that would not be recognized as an Integral needing to be factored, thus keeping the temporary `t_` in place.

File: sympy/physics/continuum_mechanics/tests/test_beam.py [hunk 4]
```diff
@@ -504,17 +504,14 @@ def test_Beam3D():
...
-            - l*x**3*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(12*E*I*(A*G*l**2 + 12*E*I))
-            - m*x**3/(6*E*I) + q*x**4/(24*E*I)
-            + l*x*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(2*A*G*(A*G*l**2 + 12*E*I))
-            - q*x**2/(2*A*G)
-            )
+    expected_deflection = (x*(A*G*q*x**3/4 + A*G*x**2*(-l*(A*G*l*(l*q - 2*m) +
+        12*E*I*q)/(A*G*l**2 + 12*E*I)/2 - m) + 3*E*I*l*(A*G*l*(l*q - 2*m) +
+        12*E*I*q)/(A*G*l**2 + 12*E*I) + x*(-A*G*l**2*q/2 +
+        3*A*G*l**2*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(A*G*l**2 + 12*E*I)/4 +
+        3*A*G*l*m/2 - 3*E*I*q))/(6*A*E*G*I))
```
Comment (smichr) [inline sympy/physics/continuum_mechanics/tests/test_beam.py]:
```python
>>> old=S('''(-l**2*q*x**2/(12*E*I) + l**2*x**2*(A*G*l*(l*q - 2*m)
...             + 12*E*I*q)/(8*E*I*(A*G*l**2 + 12*E*I)) + l*m*x**2/(4*E*I)
...             - l*x**3*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(12*E*I*(A*G*l**2 + 12*E*I))
...             - m*x**3/(6*E*I) + q*x**4/(24*E*I)
...             + l*x*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(2*A*G*(A*G*l**2 + 12*E*I))
...             - q*x**2/(2*A*G)
...             )''')
>>> new=S('''(x*(A*G*q*x**3/4 + A*G*x**2*(-l*(A*G*l*(l*q - 2*m) +
...         12*E*I*q)/(A*G*l**2 + 12*E*I)/2 - m) + 3*E*I*l*(A*G*l*(l*q - 2*m) +
...         12*E*I*q)/(A*G*l**2 + 12*E*I) + x*(-A*G*l**2*q/2 +
...         3*A*G*l**2*(A*G*l*(l*q - 2*m) + 12*E*I*q)/(A*G*l**2 + 12*E*I)/4 +
...         3*A*G*l*m/2 - 3*E*I*q))/(6*A*E*G*I))''')
>>> old.equals(new)
True
```

File: sympy/physics/continuum_mechanics/tests/test_beam.py [hunk 5]
```diff
@@ -525,12 +522,12 @@ def test_Beam3D():
     assert b2.reaction_loads == {R1: -750, R2: -750}
 
     b2.solve_slope_deflection()
-    assert b2.slope() == [0, 0, 25*x**3/(3*E*I) - 375*x**2/(E*I) + 3750*x/(E*I)]
-    expected_deflection = (25*x**4/(12*E*I) - 125*x**3/(E*I) + 1875*x**2/(E*I)
-                        - 25*x**2/(A*G) + 750*x/(A*G))
+    assert b2.slope() == [0, 0, x**2*(50*x - 2250)/(6*E*I) + 3750*x/(E*I)]
```
Comment (smichr) [inline sympy/physics/continuum_mechanics/tests/test_beam.py]:
```python
>>> S('''25*x**3/(3*E*I) - 375*x**2/(E*I) + 3750*x/(E*I)''').equals(S('''x**2*(50*x - 2250)/(6*E*I) + 3750*x/(E*I)'''))
True
```

File: sympy/physics/continuum_mechanics/tests/test_beam.py [hunk 6]
```diff
@@ -525,12 +522,12 @@ def test_Beam3D():
...
+    expected_deflection = (x*(25*A*G*x**3/2 - 750*A*G*x**2 + 4500*E*I +
+        15*x*(750*A*G - 10*E*I))/(6*A*E*G*I))
```
Comment (smichr) [inline sympy/physics/continuum_mechanics/tests/test_beam.py]:
```python
>>> old=S('''(25*x**4/(12*E*I) - 125*x**3/(E*I) + 1875*x**2/(E*I)
...                         - 25*x**2/(A*G) + 750*x/(A*G))''')
>>> new=S('''(x*(25*A*G*x**3/2 - 750*A*G*x**2 + 4500*E*I +
...         15*x*(750*A*G - 10*E*I))/(6*A*E*G*I))''')
>>> old.equals(new)
True
```

File: sympy/simplify/simplify.py [hunk 7]
```diff
@@ -591,7 +592,11 @@ def shorter(*choices):
         expr = combsimp(expr)
 
     if expr.has(Sum):
-        expr = sum_simplify(expr)
+        expr = sum_simplify(expr, **kwargs)
+
+    if expr.has(Integral):
+        expr = expr.xreplace(dict([
+            (i, factor_terms(i)) for i in expr.atoms(Integral)]))
```
Comment (jksuom) [inline sympy/simplify/simplify.py]: How does `factor_terms` transform an `Integral`?

Comment (smichr) [reply sympy/simplify/simplify.py]:
```python
>>> eq = Integral(temp, x)
>>> factor_terms(eq)
temp*Integral(1, x)
```

File: sympy/core/exprtools.py [hunk 8]
```diff
@@ -1098,6 +1098,25 @@ def handle(a):
     return terms.func(*[handle(i) for i in terms.args])
 
 
+def _factor_sum_int(expr, **kwargs):
+    from sympy.core.exprtools import factor_terms
```
Comment (jksuom) [inline sympy/core/exprtools.py]: Is this the function right below?

Comment (smichr) [reply sympy/core/exprtools.py]: ahh...leftover from when the function wasn't in the same file. I will remove that import.

File: sympy/simplify/simplify.py [hunk 9]
```diff
@@ -712,37 +703,32 @@ def sum_combine(s_t):
...
+    >>> factor_sum(s.function, s.limits)
+    y*Sum(x, (x, 1, 3))
     """
-    from sympy.core.exprtools import factor_terms
+    # XXX deprecate in favor of direct call to factor_terms
     from sympy.concrete.summations import Sum
+    kwargs = dict(radical=radical, clear=clear,
+        fraction=fraction, sign=sign)
+    expr = Sum(self, *limits) if limits else self
+    return factor_terms(expr, **kwargs)
```
Comment (jksuom) [inline sympy/simplify/simplify.py]: Should this be imported?

Comment (smichr) [reply sympy/simplify/simplify.py]: It is already there:

from sympy.core import (Basic, S, Add, Mul, Pow, Symbol, sympify, expand_mul,
                        expand_func, Function, Dummy, Expr, **factor_terms**,
                        expand_power_exp)

**Comment by smichr** [comment]
@jksuom , the thing that bothers me is that I don't feel like simplification -- at least factoring -- should separate the Sum/Integral into additive terms. Maybe that should be part of `expand_basic` for those objects. Do you have any sentiments in this regard?

**Comment by jksuom** [comment]
I guess you are referring to unevaluated Sums/Integrals. In those cases, I would expect that if simplification creates a sum of several terms, at most one of them would be of the unevaluated type.

**Comment by smichr** [comment]
> at most one of them would be of the unevaluated type.

Yes. The others can be trivially evaluated. So if `simplify` is being called I would expect that any Integral or Sum whose function is 1 should be evaluated (e.g. `Sum(1, (x, 1, y)) -> y`). I guess the question is, "Who should do what?"
```
Sum(x + y, (x, 1, 3)) -> y*Sum(1, (x, 1, 3)) + Sum(x, (x, 1, 3)) -> 3*y + Sum(x, (x, 1, 3)) -> 6 + 3*y
Sum(x*y), (x, 1, 3)) -> y*Sum(x, (x, 1, 3)) -> same -> 6*y
```
My suspicion now is that `expand_basic` should do the first, `simplify` should do the second, and `doit` should do the 3rd. `factor_terms` should only use `expand_basic` to prep an expression so things like y + Sum(y*x, (x, 1, 3)) could be factored as `y*(1 + Sum(x, (x, 1, 3))`. And perhaps a keyword for `expr_with_limits` should be added.

Products could also be handled as
```
Product(x*y,(x, 1, n)) -> Product(y, (x, 1, n))*Product(x, (x, 1, n)) -> y**n*Product(x, (x, 1, n))->y**n*n!
```

**Comment by oscarbenjamin** [comment]
+1 from me. Are you done with this?

**Comment by smichr** [comment]
It's convoluted and I made some suggestions above in the unresolved comments but those can be handled as the need arises.

**Pull request merged**